src/transformers/models/bart/modeling_bart.py

```diff
@@ -1238,9 +1356,18 @@ def forward(
         if inputs_embeds is None:
             inputs_embeds = self.embed_tokens(input) * self.embed_scale
 
-        if getattr(self.config, "_flash_attn_2_enabled", False):
+        if self._use_flash_attention_2:
             # 2d mask is passed through the layers
             attention_mask = attention_mask if (attention_mask is not None and 0 in attention_mask) else None
+        elif self._use_sdpa and not output_attentions and cross_attn_head_mask is None:
+            # output_attentions=True & cross_attn_head_mask can not be supported when using SDPA, and we fall back on
+            # the manual implementation that requires a 4D causal mask in all cases.
+            attention_mask = _prepare_4d_causal_attention_mask_for_sdpa(
+                attention_mask,
+                input_shape,
+                inputs_embeds,
+                past_key_values_length,
+            )
```

Comment:
a way to enforce same I/O is to do something like `mask_converter = _prepare_4d_causal_attention_mask_for_sdpa` here and `_prepare_4d_causal_attention_mask` below and always call mask_converter on the inputs. Nit and up to you 😉